File: csharp/rational-numbers/RationalNumbers.cs
using System;

public static class RealNumberExtension
{
    public static double Expreal(this int realNumber, RationalNumber r)
    {
        return Math.Pow(realNumber, (double)r.a / r.b);
    }
}

public struct RationalNumber
{
    public int a;
    public int b;
    public RationalNumber(int numerator, int denominator)
    {
        a = numerator;
        b = denominator;
    }

    public static RationalNumber operator +(RationalNumber r1, RationalNumber r2)
    {
        RationalNumber r = new RationalNumber(r1.a * r2.b + r2.a * r1.b, r1.b * r2.b);
        return r.Reduce();
    }

    public static RationalNumber operator -(RationalNumber r1, RationalNumber r2)
    {
        RationalNumber r = new RationalNumber(r1.a * r2.b - r2.a * r1.b, r1.b * r2.b);
        return r.Reduce();
    }

    public static RationalNumber operator *(RationalNumber r1, RationalNumber r2)
    {
        return new RationalNumber(r1.a * r2.a, r1.b * r2.b).Reduce();
    }

    public static RationalNumber operator /(RationalNumber r1, RationalNumber r2)
    {
        if (r2.a * r1.b == 0) throw new ArgumentException("bad");
        RationalNumber r = new RationalNumber(r1.a * r2.b, r2.a * r1.b).Reduce();
        return r.Reduce();
    }

    public RationalNumber Abs()
    {
        return new RationalNumber(Math.Abs(a), Math.Abs(b));
    }

    private int GCD(int a, int b) {
        while (a != 0 && b != 0) {
            if (a > b) {
                a %= b;
            } else {
                b %= a;
            }
        }
        return a == 0 ? b : a;
    }

    public RationalNumber Reduce()
    {
        int numerator = a;
        int denominator = b;
        if (numerator == 0) {
            denominator = 1;
        }
        if (denominator < 0) {
            numerator = -numerator;
            denominator = -denominator;
        }
        int gcd = GCD(Math.Abs(numerator), Math.Abs(denominator));
        return new RationalNumber(numerator / gcd, denominator / gcd);
    }

    public RationalNumber Exprational(int power)
    {
        return new RationalNumber((int)Math.Pow(a, power), (int)Math.Pow(b, power));
    }

    public double Expreal(int baseNumber)
    {
        throw new NotImplementedException("You need to implement this function.");
    }
}

File: c/rational-numbers/src/rational_numbers.c
#include "rational_numbers.h"
#include <stdlib.h>
#include <math.h>

static int get_gcd(int a, int b) {
  for (int i = abs(a); i >= 1; i--) {
    if (a % i == 0 && b % i == 0) {
      return i;
    }
  }
  return 1;
}

rational_t reduce(rational_t r) {
  if (r.numerator == 0) {
    return (rational_t) { 0, 1 };
  }
  if (r.denominator < 0) {
    r.numerator *= -1;
    r.denominator *= -1;
  }
  int gcd = get_gcd(r.numerator, r.denominator);
  return (rational_t) {
    r.numerator / gcd,
    r.denominator / gcd
  };
}

rational_t add(rational_t r1, rational_t r2) {
  return reduce((rational_t) {
    r1.numerator * r2.denominator + r2.numerator * r1.denominator,
    r1.denominator * r2.denominator
  });
}

rational_t subtract(rational_t r1, rational_t r2) {
  return reduce((rational_t) {
    r1.numerator * r2.denominator - r2.numerator * r1.denominator,
    r1.denominator * r2.denominator
  });
}

rational_t multiply(rational_t r1, rational_t r2) {
  return reduce((rational_t) {
    r1.numerator * r2.numerator,
    r1.denominator * r2.denominator
  });
}

rational_t divide(rational_t r1, rational_t r2) {
  return reduce((rational_t) {
    r1.numerator * r2.denominator,
    r1.denominator * r2.numerator
  });
}

rational_t absolute(rational_t r) {
  return reduce((rational_t) {
    abs(r.numerator),
    abs(r.denominator)
  });
}

rational_t exp_rational(rational_t r, int n) {
  return reduce((rational_t) {
    pow(r.numerator, n),
    pow(r.denominator, n)
  });
}

float exp_real(int x, rational_t r) {
  return pow(pow(x, r.numerator), (1 / (float)r.denominator));
}
